devtools/src/arcs-hcr-list.js

```diff
@@ -0,0 +1,314 @@
+      switch (msg.messageType) {
+        case 'arc-available': {
+          if (!msg.messageBody.speculative) {
+            const id = msg.meta.arcId.toString();
+            this.arcIdsToKeepFresh.push(id);
+          }
+          break;
+        }
+        case 'watch-particle-sources': {
+          const before = this.particleSources.size;
+          for (const particleSource of msg.messageBody) {
+            this.particleSources.add(particleSource);
+          }
+          if (this.particleSources.size > before && this.websocket != null) {
+            const files = JSON.stringify([...this.particleSources]);
+            this.websocket.send(files);
+          } 
+          break;
+        }
+      }
+    }
+  }
+
+  onMessageBundle(messages) {
+    for (const msg of messages) {
+      switch (msg.messageType) {
+        case 'arc-selected':
+          this.reloadableParticles = [];
+          this.allParticlesInArc = [];
+          break;
+
+        case 'PecLog': {
+          const m = msg.messageBody;
+          switch (m.name) {
+            case 'InstantiateParticle': {
+              const particleId = m.pecMsgBody.id.toString();
+              const spec = m.pecMsgBody.spec;
+
+              this.allParticlesInArc.push({id: particleId, spec: spec});
+              this._displayLog(particleId, spec);
+            }
+          }
+          break;
+        }
+      }
+    }
+  }
+
+  _forceReload() {
+    this.reloadableParticles.forEach(particle => {
+      this._updateLog(particle.implFile);
+      this.arcIdsToKeepFresh.forEach(id => {
+        this.send({
+            messageType: 'particle-reload',
+            messageBody: particle.implFile,
+            arcId: id
+        });
+      });
+    });
+  }
+
+  _displayLog(particleId, spec) {
+    if (this.pathToTimeLog.has(spec.implFile)) {
+      const particle = {
+        name: spec.name,
+        id: particleId,
+        implFile: spec.implFile,
+        time: this.pathToTimeLog.get(spec.implFile)
+      };
+      this.reloadableParticles.push(particle);
+    }
+  }
+
+  _updateLog(filePath) {
+    const time = this._getCurrentTime();
+    if (!this.pathToTimeLog.has(filePath)) {
+      this.pathToTimeLog.set(filePath, []);
+    }
+    const timeLog = this.pathToTimeLog.get(filePath);
+    if (!timeLog.includes(time)) timeLog.push(time);
+
+    [...this.reloadableParticles].forEach(particle => {
+      if (particle.implFile == filePath) {
+        const explorer =  this.shadowRoot.querySelector(`object-explorer[log-id="${filePath}"]`);
+        if (explorer) {
+          explorer.refresh();
+          explorer.flash();
+        }
+      }
+    });
+  }
+
+  _loadServer() {
+    if (this.$.toggle.checked) {
+      this.$.toggle.checked = false;
+
+      this.websocket = new WebSocket('ws://localhost:8888');
+      this.websocket.onopen = e => {
+        this.$.toggle.checked = true;
+
+        this.websocket.onmessage = msg => {
+          const input = (`${msg.data}`).split(' ');
```

Comment:
Why do you need to do `${msg.data}` ? Isn't msg.data a string?

Maybe this would be more readable:
`const [operation, path] = msg.data.split(' ');`

Then you don't have to write `input[0]` or `input[1]` which are obscure

You could also consider serializing a deserializing a JSON message, though I'm ok with this space concatenated messages for such a simple case.

Reply:
Changed using JSON serialize & deserialize